### PR TITLE
[Merged by Bors] - UCB: add standalone proposal db

### LINKED ISSUE #2669: Clean up terminology: "verify" vs. "validate"
## Description
We currently conflate the terms "verify" and "validate" in several places, such as in the mesh, and use them interchangeably, when we refer to passing a layer to the tortoise for processing:

https://github.com/spacemeshos/go-spacemesh/blob/9d3a379994e3652eb0378d2cc555588bba91005e/mesh/mesh.go#L350-L355

Clean this up throughout the documentation and codebase.

### PR DESCRIPTION
**Pull request by countvonzero**
## Motivation

2nd PR for #2906
Closes #2669

## Changes

- add proposal db outside of mesh to prepare for next PR
- update some mesh method names to reflect recent code change 

## Test Plan

unit test

## DevOps Notes

- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

### CODE REVIEW

File: layerfetcher/layers_test.go [hunk 1]
```diff
@@ -270,7 +270,7 @@ func generateLayerBlocksWithHash(consistentHash bool, numInputVector int) []byte
 	}
 	lb := layerBlocks{
 		Blocks:         blockIDs,
-		InputVector:    iv,
+		HareOutput:     iv,
```
Comment (nkryuchkov) [inline layerfetcher/layers_test.go]: I'd also rename `iv`. E.g. to `ho`

File: mesh/mesh_test.go [hunk 2]
```diff
@@ -519,10 +519,10 @@ func TestMesh_HandleValidatedLayer(t *testing.T) {
 	tm.mockState.EXPECT().GetStateRoot().Return(types.Hash32{}).Times(1)
 	tm.mockState.EXPECT().ValidateNonceAndBalance(gomock.Any()).Return(nil).AnyTimes()
 	tm.mockState.EXPECT().AddTxToPool(gomock.Any()).Return(nil).AnyTimes()
-	tm.HandleValidatedLayer(context.TODO(), gPlus2, blockIDs)
+	tm.ProcessLayerPerHareOutput(context.TODO(), gPlus2, blockIDs)
 
 	// input vector saved
-	iv, err := tm.GetLayerInputVectorByID(gPlus2)
+	iv, err := tm.GetHareConsensusOutput(gPlus2)
```
Comment (nkryuchkov) [inline mesh/mesh_test.go]: I'd rename `iv` and update the comment

File: mesh/mesh_test.go [hunk 3]
```diff
@@ -547,10 +547,10 @@ func TestMesh_HandleValidatedLayer_emptyOutputFromHare(t *testing.T) {
 	tm.mockState.EXPECT().GetStateRoot().Return(types.Hash32{}).Times(1)
 	tm.mockState.EXPECT().ValidateNonceAndBalance(gomock.Any()).Return(nil).AnyTimes()
 	tm.mockState.EXPECT().AddTxToPool(gomock.Any()).Return(nil).AnyTimes()
-	tm.HandleValidatedLayer(context.TODO(), gPlus2, empty)
+	tm.ProcessLayerPerHareOutput(context.TODO(), gPlus2, empty)
 
 	// input vector saved
-	iv, err := tm.GetLayerInputVectorByID(gPlus2)
+	iv, err := tm.GetHareConsensusOutput(gPlus2)
```
Comment (nkryuchkov) [inline mesh/mesh_test.go]: `iv`

File: proposals/proposaldb.go [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+		opt(db)
+	}
+	return db
+}
+
+// NewProposalDB returns a new DB for proposals.
+func NewProposalDB(path string, msh meshDB, logger log.Log) (*DB, error) {
+	pdb, err := database.NewLDBDatabase(filepath.Join(path, "proposals"), 0, 0, logger)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize proposals db: %v", err)
```
Comment (nkryuchkov) [inline proposals/proposaldb.go]:
```suggestion
		return nil, fmt.Errorf("failed to initialize proposals db: %w", err)
```

File: proposals/proposaldb.go [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+	}
+	ldb, err := database.NewLDBDatabase(filepath.Join(path, "proposal_layers"), 0, 0, logger)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize proposal layers db: %v", err)
```
Comment (nkryuchkov) [inline proposals/proposaldb.go]:
```suggestion
		return nil, fmt.Errorf("failed to initialize proposal layers db: %w", err)
```

File: proposals/proposaldb.go [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+	defer db.mu.Unlock()
+	dbp := &types.DBProposal{
+		ID:         p.ID(),
+		BallotID:   p.Ballot.ID(),
+		LayerIndex: p.LayerIndex,
+		TxIDs:      p.TxIDs,
+		Signature:  p.Signature,
+	}
+	if data, err := codec.Encode(dbp); err != nil {
+		return fmt.Errorf("could not encode dbproposal: %w", err)
```
Comment (nkryuchkov) [inline proposals/proposaldb.go]:
```suggestion
		return fmt.Errorf("could not encode db proposal: %w", err)
```

File: tortoise/tortoise.go [hunk 7]
```diff
@@ -717,7 +717,7 @@ func (t *turtle) addLocalVotes(ctx context.Context, logger log.Log, lid types.La
 
 	if !lid.Before(t.layerCutoff()) {
 		// for newer layers, we vote according to the local opinion (input vector, from hare or sync)
-		opinionVec, err := t.bdp.GetLayerInputVectorByID(lid)
+		opinionVec, err := t.bdp.GetHareConsensusOutput(lid)
```
Comment (nkryuchkov) [inline tortoise/tortoise.go]: Should `opinionVec` be renamed?

**Review by nkryuchkov — CHANGES_REQUESTED** [review]

**Review by dshulyak — APPROVED** [review]
it's a pity that we are stuck with this non-sensical database per namespace pattern, as it leads to more code that has to be rewritten regardless of what database approach will be used in https://github.com/spacemeshos/go-spacemesh/issues/2918

File: layerfetcher/layers_test.go (hunk 1 above)
Comment (countvonzero) [reply layerfetcher/layers_test.go]: done

**Comment by countvonzero** [comment]
> it's a pity that we are stuck with this non-sensical database per namespace pattern, as it leads to more code that has to be rewritten regardless of what database approach will be used in #2918

for me the code rewrite is less of a problem than the lack of atomicity between writes caused by separate databases.

**Comment by countvonzero** [comment]
@nkryuchkov i've done a more extensive rename. PTAL. thanks

File: proposals/proposaldb.go [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+// Close closes all resources.
+func (db *DB) Close() {
+	db.proposals.Close()
+	db.layers.Close()
+}
+
+// HasProposal returns true if the database has the Proposal specified by the ProposalID and false otherwise.
+func (db *DB) HasProposal(id types.ProposalID) bool {
+	has, err := db.proposals.Has(id.Bytes())
+	return err == nil && has
```
Comment (nkryuchkov) [inline proposals/proposaldb.go:87]: I think it's not a big deal but on DB error other than the not-found error we will not know that it occurred.

**Review by nkryuchkov — APPROVED** [review]

**Comment by nkryuchkov** [comment]
@countvonzero thanks

File: proposals/proposaldb.go (hunk 8 above)
Comment (countvonzero) [reply proposals/proposaldb.go:87]: this is used in proposal handler. my consideration was, either kind of error (not found or disk error), the resulting action is similar. we still want to process the newly received proposal and save it as best effort. so checking the error wouldn't be very meaningful, and whatever disk error will recur later when the new proposal is saved.

there is similar logic for Ballot.

Comment (nkryuchkov) [reply proposals/proposaldb.go:87]: I agree that such an error shouldn't change the handling process. My thought was to consider logging a disk error to know that it happened. But it's not important, so we can also leave it as is

**Comment by countvonzero** [comment]
bors merge

**Comment by countvonzero** [comment]
bors merge

**Comment by countvonzero** [comment]
bors merge

**Comment by countvonzero** [comment]
bors merge

**Comment by countvonzero** [comment]
bors merge

**Pull request closed without merging**